Accelerate WTF::equal using NEON intrinsics

ddegazio opened the pull request:
#### 661be0f9c16d3bdefaf9c80e29eb793428440dc7
<pre>
Accelerate WTF::equal using NEON intrinsics
<a href="https://bugs.webkit.org/show_bug.cgi?id=263734">https://bugs.webkit.org/show_bug.cgi?id=263734</a>
rdar://117541948

Reviewed by Yusuke Suzuki.

1. Adds a hash check to WTF::equal(StringImpl*, StringImpl*) to quickly rule
out matches (thanks @hyjorc1!).

2. Reworks the implementation of WTF::equal for LChar and UChar strings.
Instead of checking for leftover bytes at the end, we use overlapping loads to
reduce the amount of branching per string. And we now switch up-front on the
string length instead of falling through to other branches, reducing branching
for most small strings.

3. When CPU(ARM_NEON) is defined, use 16-byte SIMD chunks to do the comparison,
whenever strings are large enough.

4. Also adds a SIMD loop to WTF::equal(const LChar*, const UChar*, unsigned),
which is dramatically (~5x) faster than the current byte-by-byte pessimistic
implementation.

* Source/WTF/wtf/PlatformCPU.h:
* Source/WTF/wtf/text/StringCommon.h:
(WTF::equal):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::equal):
* Source/WTF/wtf/text/StringImpl.h:

Canonical link: <a href="https://commits.webkit.org/269866@main">https://commits.webkit.org/269866@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ea240c26aeecdab5d7a6b52a0aa2aae3226b2fd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23728 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25881 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21898 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24250 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22443 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23971 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1438 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20529 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26477 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1213 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21435 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27703 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20647 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21686 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21709 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25472 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/23062 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1137 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18832 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/30455 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1127 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6690 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5706 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1549 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/30409 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1438 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6359 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -1458,6 +1458,8 @@ bool equal(const StringImpl* a, const LChar* b)
 
 bool equal(const StringImpl& a, const StringImpl& b)
 {
+    if (a.hash() != b.hash())
```
If you check the implementation of StringImpl's `hash`, it says

```
unsigned hash() const { return hasHash() ? rawHash() : hashSlowCase(); }
```

So, for comparing strings with large sizes, maybe we should check and compare `rawHash()` (the cached hash value) since `hashSlowCase()` (the hash computation) might be costly.

hyjorc1 replied:
Something like this?

```
if (a.hasHash() && b.hasHash()) {
    if (a.rawHash() != b.rawHash())
        return false;
}
```

Constellation replied:
Yes.
1. StringImpl::hash computes the hash and store it, in a non-thread safe way. So we should not always do it in underlying function (DFG may use the other thread's StringImpl only for comparison).
2. hashSlowCase is costly.

Let's do this,
```
unsigned aHash = a.rawHash();
unsigned bHash = b.rawHash();
if (aHash != bHash && aHash && bHash)
    return false;
```

ddegazio replied:
Ah, thank you for pointing this out! I still have to A/B test this particular change so it fell under the radar, I'll fix it right away.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -70,72 +71,107 @@ bool equalLettersIgnoringASCIICase(const char*, ASCIILiteral);
+        return true;
+#else
+    default: // Length is longer than 8 (16 bytes).
+        static_assert(false);
```
We are hitting this in GTK ports etc.

ddegazio replied:
Oh whoops, that absolutely should not still be in, it was just to confirm that `CPU(ARM_NEON)` was defined. 🤦

ddegazio replied:
Removed in latest push.

Constellation reviewed (approved):
r=me with comments.

ddegazio committed:
Accelerate WTF::equal using NEON intrinsics
https://bugs.webkit.org/show_bug.cgi?id=263734
rdar://117541948

Reviewed by Yusuke Suzuki.

1. Adds a hash check to WTF::equal(StringImpl*, StringImpl*) to quickly rule
out matches (thanks @hyjorc1!).

2. Reworks the implementation of WTF::equal for LChar and UChar strings.
Instead of checking for leftover bytes at the end, we use overlapping loads to
reduce the amount of branching per string. And we now switch up-front on the
string length instead of falling through to other branches, reducing branching
for most small strings.

3. When CPU(ARM_NEON) is defined, use 16-byte SIMD chunks to do the comparison,
whenever strings are large enough.

4. Also adds a SIMD loop to WTF::equal(const LChar*, const UChar*, unsigned),
which is dramatically (~5x) faster than the current byte-by-byte pessimistic
implementation.

* Source/WTF/wtf/PlatformCPU.h:
* Source/WTF/wtf/text/StringCommon.h:
(WTF::equal):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::equal):
* Source/WTF/wtf/text/StringImpl.h:

Canonical link: https://commits.webkit.org/269866@main
